### DISCUSSION TITLE
Issues when running airflow on AKS with azure sql as well as Azure Database for PostgreSQL Single Server

### DISCUSSION THREAD

**Discussion by Raul824** [apache/airflow#31405] [Q&A] [answered]
We are trying to setup airflow on AKS but we are facing multiple issues with multiple different database back-end setups.

1. Setting up back-end with azure sql server
When we setup database back-end as azure sql server after putting more than 500 dags we started getting communication link failure from sql server odbc driver and no new dags are appearing in GUI and some of the dags which appear are missing from serialized_dag table and the error is thrown dag is missing from dagbag when trying to view the dag. 

2. Setting up back-end with Azure Database for PostgreSQL Single Server
When we setup database back-end as Azure Database for PostgreSQL Single Server after putting more than 500 dags we get no errors in scheduler but in azurediagnostic "could not receive data from client: An existing connection was forcibly closed by the remote host." no new dags are appearing in GUI and some of the dags which appear are missing from serialized_dag table and the error is thrown dag is missing from dagbag when trying to view the dag. 

3. Running postgres as back-end inside a pod with and without pgbouncer.
all 1500 dags parse within 5 minutes but when we start running more than 100 tasks the webui becomes unresponsive for sometime but everything still runs.

If anyone has setup Airflow on aks could you please share you configuration and how many task concurrency and worker are you running.

**Comment by Raul824** [comment]
This is the error we are getting in scheduler logs.


[2023-05-19T05:16:19.344+0000] {manager.py:306} ERROR - DagFileProcessorManager (PID=10611) last sent a heartbeat 50.99 seconds ago! Restarting it
[2023-05-19T05:16:19.346+0000] {process_utils.py:135} INFO - Sending Signals.SIGTERM to group 10611. PIDs of all processes in the group: [12533, 12534, 10611]
[2023-05-19T05:16:19.346+0000] {process_utils.py:86} INFO - Sending the signal Signals.SIGTERM to group 10611
[2023-05-19T05:16:20.434+0000] {process_utils.py:258} INFO - Waiting up to 5 seconds for processes to exit...
[2023-05-19T05:16:20.471+0000] {process_utils.py:79} INFO - Process psutil.Process(pid=12534, status='terminated', started='05:16:18') (12534) terminated with exit code None
[2023-05-19T05:16:20.473+0000] {process_utils.py:79} INFO - Process psutil.Process(pid=12533, status='terminated', started='05:16:18') (12533) terminated with exit code None
Process ForkProcess-4:
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/usr/local/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/dag_processing/manager.py", line 259, in _run_processor_manager
    processor_manager.start()
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/dag_processing/manager.py", line 493, in start
    return self._run_parsing_loop()
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/dag_processing/manager.py", line 616, in _run_parsing_loop
    self._collect_results_from_processor(processor)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/dag_processing/manager.py", line 1020, in _collect_results_from_processor
    if processor.result is not None:
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/dag_processing/processor.py", line 343, in result
    if not self.done:
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/dag_processing/processor.py", line 311, in done
    if self._parent_channel.poll():
  File "/usr/local/lib/python3.7/multiprocessing/connection.py", line 255, in poll
    self._check_closed()
  File "/usr/local/lib/python3.7/multiprocessing/connection.py", line 136, in _check_closed
    raise OSError("handle is closed")
OSError: handle is closed
[2023-05-19T05:16:20.566+0000] {process_utils.py:79} INFO - Process psutil.Process(pid=10611, status='terminated', exitcode=1, started='05:00:27') (10611) terminated with exit code 1

**Comment by potiuk** [comment]
I think you should tell which version of airlow you are running, what configuration you  have (which executor) and more importantly what kind of distribution mechanism you use for files.

Generally the suggestion is to always use latest released version of Airlfow - because with every version we find and fix many problem and what you experience might be result of some of them (but it's hard to say without all the dettails, explaining where exactly the problem happens and seeing the stack traces (and knowing the above version and configuration information). 

There might be a number things that you can fine tune

1)  If you get ODBC communication errors  in scheduler - you did not explain where you get the errors, so I am guessing - (especially if you are trying to use MSSQL or MYSQL) you might try to limit the default https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#max-tis-per-query  (but best is to check on your server what is the reason for those errors) - and generally you might read through https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/scheduler.html#fine-tuning-your-scheduler-performance. to fine-tune some of the parameters

Also you might simply have not enough memory or other resources for your database, so you shoudl look at the logs of your database and your monitoring to see if whatever resources you have (memory, filesize, CPU etc. is not getting exhausted.

2. Similarly for that one - it's hard to say but likely resources is the culprit. You can also configure the above limit (though postgres should be much safer) - but maybe there are some configuration settings in the server of Azure that limit Postgres. Generally speaking ALWAYS connect via PgBouncer with Postgres when you run Airflow. Airflow opens many connections and in Postgres model, it will open a new process for every connection and allocate memory , so using PGBouncer is the only approach that gives you scalability.

3. This is the settings that you should never use in any scale. Running Postgres and PGBouncer in POD (generally statefully in Kubernetes) is not a good idea and in our chart it is purely a development setting (not sure which chart you are using and whether you use any chart). When speaking of unresponsive UI - it's hard to say - it could be memory again, it also could be you have not enough CPU allocated for your PODs. Generally it looks like resource issue.

**Reply by Raul824** [reply]
thanks a lot for details.

Airflow version 2.6.0 will update to 2.6.1. Have tested same with airflow 2.5.
We couldn't use any helm chart because we don't have public internet access in our cloud services.

The Communication link failures start coming as soon as we put  700+ dags without even enabling them.
We tried putting them in batches  but as soon as that count reaches 700-800+ we start getting these errors.

We tried increasing the database cpu and memory but as soon as the database is deployed as a service we are seeing issues.

I know that postgres as a pod is not recommended but I was trying to do a POC to prove it's not airflow or postgres database(As delivering this was some other teams responsibility which kept on saying issues in airflow and aks). And I just wanted to show them it might be the database service which might be the issue.

If someone can let me know the minimum recommended specification for AKS cluster size and database service specification for running airflow that can handle 200 concurrent tasks at a time and 1000+ dags. 
I can put those requirements clearly and point out the issues.


The main issue is am new to airflow and Microsoft keeps on pushing the issues to airflow as open source and no support available. They tried pushing ADF pipelines on to me several times which I rejected and insisted on airflow as it was running fine on an on-premise machine with 128gigs of ram and 16 cpu cores which we are now planning to scale if AKS solution doesn't work as planned.

**Reply by potiuk** [reply]
> If someone can let me know the minimum recommended specification for AKS cluster size and database service specification for running airflow that can handle 200 concurrent tasks at a time and 1000+ dags.

It's really impossible and depends what your DAGs are doing - even importing the dags if they are not running will take resources. If you start getting problems with big number of DAGs that you have not even enabled, then it might indicate that you need to improve the way your DAGs are written and the problem comes from your code. Youshould take a look at best practices of Airflow. https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#top-level-python-code - you are likely to use Variables, Connections or perform other database-related operations whiile your DAGs are being parsed (and you should not).

> The main issue is am new to airflow and Microsoft keeps on pushing the issues to airflow as open source and no support available. They tried pushing ADF pipelines on to me several times which I rejected and insisted on airflow as it was running fine on an on-premise machine with 128gigs of ram and 16 cpu cores which we are now planning to scale if AKS solution doesn't work as planned.

Well Actually the ONLY thing you pay extra for with ADF Microsoft service is the support they should provide. We - in the open source do not provide support - becuase this is the open source project. If you are unhappy with them - there are other providers who not only cooperate with Airflow and have their own contributors and work on getting committers, but they also do not push problems on open-source team - where they are providing you the "service" that should take care about deployment and scaling. 

If they don't give you the service you need, consider switching it to any of the competition. If you can't switch off Azure, I recommend you look at Astronomer. They provide great support for Airflow on Azure as well.

**Reply by ashb** [reply, marked as answer]
Hi @Raul824 Yeah, we at Astronomer have plenty of customers running our Airflow stack on Azure (not using ADF! And unlike Microsoft we have a team of people who know Airflow and can _actually_ support it -- I'm on that rota too) so if you want to chat more about it email me and I can hook you up with a demo (and can be as sales-y or not as you like, I'm an engineer and I won't give your details to sales people without your approval) -- email address is in my bio https://github.com/ashb.
